File: java/src/test/java/pl/kkurczewski/deser/value/ValueObjectControllerTest.java
package pl.kkurczewski.deser.value;

import pl.kkurczewski.deser.value.dto.ValueObject;
import pl.kkurczewski.deser.value.dto.Wrapper;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest;
import org.springframework.test.web.reactive.server.WebTestClient;

import java.net.URI;

import static org.springframework.http.MediaType.APPLICATION_JSON;

@WebFluxTest(ValueObjectController.class)
class ValueObjectControllerTest {

    @Autowired
    private WebTestClient webClient;

    @Test
    void shouldEchoWrapperRequest() {
        Wrapper wrapper = new Wrapper(ValueObject.from("foo"), ValueObject.from("bar"));

        webClient
                .post()
                .uri(URI.create("/wrapper"))
                .contentType(APPLICATION_JSON)
                .bodyValue(wrapper)
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Wrapper.class)
                .isEqualTo(wrapper)
                .returnResult();
    }

    @Test
    void validateJsonWrapperFormat() {
        Wrapper wrapper = new Wrapper(ValueObject.from("foo"), ValueObject.from("bar"));

        webClient
                .post()
                .uri(URI.create("/wrapper"))
                .contentType(APPLICATION_JSON)
                .bodyValue(wrapper)
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody()
                .json(normalize("{'first':'foo','second':'bar'}"));
    }

    private String normalize(String json) {
        return json.replaceAll("'", "\"");
    }
}

File: kotlin/src/test/kotlin/pl/kkurczewski/json/value/ValueObjectControllerTest.kt
package pl.kkurczewski.json.value

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.http.MediaType
import org.springframework.test.web.reactive.server.WebTestClient
import pl.kkurczewski.json.value.dto.ValueObject
import pl.kkurczewski.json.value.dto.Wrapper
import java.net.URI

@WebFluxTest(ValueObjectController::class)
class ValueObjectControllerTest(@Autowired val webClient: WebTestClient) {

    @Test
    fun shouldEchoWrapperRequest() {
        val wrapper = Wrapper(ValueObject("foo"), ValueObject("bar"))
        val body = webClient
                .post()
                .uri(URI.create("/wrapper"))
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(wrapper)
                .exchange()
                .expectStatus()
                .is2xxSuccessful
                .expectBody(Wrapper::class.java)
                .returnResult()
                .responseBody

        assertThat(body).isEqualTo(wrapper)
    }

    @Test
    fun validateJsonWrapperFormat() {
        val wrapper = Wrapper(ValueObject("foo"), ValueObject("bar"))
        webClient
                .post()
                .uri(URI.create("/wrapper"))
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(wrapper)
                .exchange()
                .expectStatus()
                .is2xxSuccessful
                .expectBody()
                .json(normalize("{'first':'foo','second':'bar'}"))
    }

    private fun normalize(json: String): String {
        return json.replace("'".toRegex(), "\"")
    }
}